README.md
# DEAD PROJECT!
본론에 앞서 프로젝트는 현재 종료 되었고 간간이 업데이트는 있을 예정이지만 실질적인 변화가 있지는 않을 것이다. 프로젝트 종료 이유는 첫번째는 개발자 (본인)의 실력이 너무 낮아 인공지능의 성능을 내지 못한 것이다. 두번째는 시간적인 문제였다. 현 프로젝트는 혼자 진행이 아닌 팀으로 진행이 되는데 너무 늦어지고 지체가 되는 바람에 일찍이 접었다. 문제점을 파악했거나 혹은 수정을 하여 본인이 작성하고 싶다면 언제든지 Fork를 하고 작성만 해준다면 좋다.

## ECG DBN Project
ECG Singal data를 받아 DBN 으로 각 Feature를 뽑아내는 인공지능 모델 프로젝트이다. 현 프로젝트는 아래 리뷰 논문에 <a href="https://www.sciencedirect.com/science/article/pii/S2590188520300123?via%3Dihub">GB-DBN을 활용한 논문</a>을 기반으로 진행되고 있다. 

GB-DBN은 RBM를 여러겹 쌓은 Deep Learning Machine이기에 학습을 시키기 위해서는 RBM 모델이 여러개가 필요하다. 즉, 여러개의 RBM 모델을 쌓아 DBN을 만드는 것이기에 여기서 중요한 부분은 확률 분포 모델을 어떤 것을 활용하냐이다. Bernouli 확률 분포는 가장 기본적인 RBM의 확률 분포 모델이고 GB-RBM은 Gaussian 확률 분포를 활용한 모델이다. GB-DBN도 이름에서 알 수 있듯 Gaussian 확률 분포를 활용하였기에 GB-RBM과 BB-RBM를 같이 활용하여 구성된다. 
# Code description
## Model codes
빠르게 작업을 하다보니 현재 디렉토리가 제대로 정리가 되어 있지가 않다. 아마 6월 전에는 한번 정리를 할 것 같으나 현재는 result를 남겨둬야 한다는 정메서 *.ipynb 파일들은 모두 현재 모델 코드이다. 다만 몇몇 설정 값과 모델 구조의 차이가 있으니 참고하길 바란다. 정 안된다면 CHANGES.txt 를 참고하거나 Issue 를 남기면 이에 대해 답변 하겠다. 
# Status
현재 모든 진행 상황은 CHANGES.txt에 기록을 하고 있으며 기록 후 변경 내용이 있을때마다 커밋을 하고 있다. 그렇기에 과거 지워진 내용이나 달라진 내용을 확인하고 싶다면 commit history를 보면 알 수 있다. 


CHANGES.txt
# DEADLINE!
5월 17일 까지 제작하여 accuaracy 뽑아내기
May. 5. 2022 (B. InSung)
- Linear 모델 까지 통과, 이제 정상 비트냐 이소성 비트냐만 결과 도출하면 끝

May. 22. 2022 (B. InSung)
- Project DECLINED.
    - 계속 좋지 않은 결과 물이 나와 최종적으로 프로젝트 종료를 진행하였다. 개인적인 생각으로 문제는 Pre-processing 이 GB-DBN에서는 가장 중요한데 이를 제대로 못 해준 점, 그리고 기본적으로 나의 실력이 많이 미숙한 점이 가장 크다. 이 번 기회에로 Torch를 접했으니 일단 지속적으로 공부하고 코드보다는 모델의 이해에 좀 더 집중할려고 한다.
    - Project is done. 
        May. 22. 2022. 
        PM  21:46

May. 20. 2022 (B. InSung)
- RuntimeError: CUDA error: device-side assert triggered
    - mat_reader_model.ipynb에서 오류 발생
    - svm 모델에서 X 할당에서 오류 발생. 백업 커밋.
- svm model 할당 문제 해결. 현재 트레인 진행중 일부 내용 커밋.
- 최종적으로 추출된 모델을 돌렸을때 50% 언저리의 결과 값을 나오게됨.
    - 학습 결과 값 : 82.2429906542056GBRBM

May. 19. 2022 (B. InSung)
- 아래 Todo 를 진행하였다. 문제는 확률에도 큰 변화가 없다는 것을 발견. 이를 올리기 위해서는 개인적인 생각으로는 전처리 과정이 굉장히 많이 요구 되는 것으로 보임. 이를 해결하기 위해서는 이전에 활용 했던 방법으로 해야할거 같은 느낌.
    - 개인적으로 열차내에서 진행해볼 예정.
- Late thursday.
    - 새로운 matlab 데이터 셋으로 구성된 내용을 읽어와 학습진행.
    - 현재 사이즈 매치가 되지 않아 오류 발생. 백업 커밋
    - 23:41:21 모델 Train 시작. 아침에 확인예정
      - epoch : 200
      - 187 visible units, 80, 100, 120, 150, 200, 250 (? 아무튼 논문대로) 하여 학습 진행중

May. 18. 2022 (B. InSung)
- SVM Train을 한번에 동시에 돌리자 Out of bounds 오류 발생. 이를 해결하기 위해 디버깅 진행 중.
    - 현재 작업 파일 명: cuda_model_csv_cheese.ipynb
- 말이 안됨; 100%가 나왔다.
- cuda_model_csv_cheese.ipyn에서 새로운 네트워크를 구성하는 중.
    - cuda_model_csv_cheese.ipynb 사망함.
- cuda_model_csv_cheese copy.ipynb 에서는 기존 네트워크에서 잘 나온 모델 저장.
# TODO
    // - SVM Input 이전 표준화.
    // - Normal을 제외, q 제외, 학습하기.
    - EPOCH = 200 으로 늘리기
    - Visible Layer 의 갯수 187
    - 180, 80, 200 BB-DBN Hidden Layer의 갯수
    - 100, 250, 120 GB-DBN Hidden Layer의 유닛 갯수.

May. 17. 2022 (B. InSung)
- Late Tuesday Session
- SVM Linear 학습 진행은 완료됨. 다만 문제는 accuaracy 확인 불가. 일단 테스트 까지 하고 결과를 볼 예정.

May. 15. 2022 (B. InSung)
- Sunday session is now on.
- 일단 학습 정확율이 90% 까지는 올라왔는데 이게 맞나? 아니 근데 갑자기 이렇게 된것도 이상하고 90.0, 80.0 이렇게 딱 떨이지는 것도 이상한거 같다.
- 본인 허리가 지금 너무 아픔. 일단 백업 커밋을 올린다만, 현재까지 진행 사항
    - gb dbn 에서 나온 값들을 정리하고 의미하는 바 작성하여 보고서 적기.
    - SVM 내에서 나온 값을 그래프와 시키고 시각화 진행 중에 train 함수와 plot 함수를 merge 진행

May. 14. 2022 (B. InSung)
- 아ㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏ 졸려요. 배고프고.
- 일단 모델은 돌아가는거 같은데 cuda 로 돌리고 있다. 유감스럽게도 cpu로 하면 한 비트당 50초 씩 걸린다...
아무튼 일단 전체 bbrbm, gbrbm 돌리고 결과값과 함께 커밋 예정. (1:17am)
- 정말 드는 의문점 하나. 아니 왜 쿠다로 돌리는데 cpu가 제일 바쁘고 gpu 리소스가 왜 0이지?
- loss가 50을 넘지를 못한다. bbrbm 이후 gbrbm에서의 변화를 기대...
- 아침이 밝았습니다... 아침입니다 열어분... 지금 일단 gbrbm 에서 넘어가는 과정 중에서 cuda 오류 발생. 이건 고치고 가야한다... 너무 많이 시간을 잡아먹는다...
================================================
- 자 현재까지의 진행 사항. 
    - 일단 지금까지 해왔던 Train이 맞는가? 일단 아닌듯. 형이랑 얘기를 해보니 전에 돌린거는 얘따로 쟤따로 놀고 있는데 기강이 풀린듯. 한번에 싹 집합 시켜서 학습시켜서 정신차리게 하는중.

May. 13. 2022 (B. InSung)
- First BBRBM은 작동 여부 확인. 이를 Second BBRBM까지 이어 갈수 있는지는 아직 확인중.
- Second BBRBM으로 변환 하는 과정 중에 확률 분포 에러 발생. 백업 커밋.
- 사이즈 오류로 인해 생기는 모델 input data 오류 제거.
    - Second BBRBM 레이어를 통과시키는 것을 이제 주 목적.
- BBRBM, GBRBM 설계 완료. 테스트 진행 예정. 커밋.

May. 12. 2022 (B. InSung)
- 내가 뭘 했는지는 몰라도 일단 뭐 되긴 된거 같다. 연산 중에 오류로 torch.mv 함수 내에서 벡터와 텐서 연산 오류 외에는 아직 큰 문제는 없음.
- torch.mv 함수 내에서 matmul 과정 중에서 오류 발생. 아마도 13000000 텐서와 10 벡터를 연산을 하는데 텐서를 10 * 1300000으로 맞춰 줘야하는거 같다.
    - .view(10, 1300000) 함수로 변환시 오류 없음. torch.mv 에서 적용 후 검토 예정
    RuntimeError: size mismatch, got 10, 10x1300000,10 
- 사이즈를 반대로 적용. 1300000 * 10 으로 적용하여 값 확인
    - 오류 해결됨. 커밋.
    ===============================================
- MSI 연구실 컴퓨터로 작업 시작.
    - 기존 데이터를 입력할때 맞지 않게 입력을 했음.
- 일단 first bbrbm은 확인. 전체 다 돌리고 결과를 볼 예정. 커밋.

May. 11. 2022 (B. InSung)
- 모델을 재구성하는중.
    - 기존에 모델의 가장 크나큰 문제는 일단 제대로 된 값이 하나도 없었던 거 같으며 라벨 처리를 따로 안함. 현재 문제도 거의 동일하여 라벨 처리를 제대로 못해주고 있긴하다만,
    지금은 조금 다른 문제에서 난관을 겪고 있음.
    - 데이터의 총 길이 값이 1,300,000 개로 이루워져 있어서 output 값을 지정할려면 1, 1300000으로 해줘야하나 메모리 할당에서 오류 발생
    - 멘탈이 나가뒤져가는데 일단 커밋은 해야할거 같다.
- Output size도 맞게 한거 같고 다 맞게 한거 같은데 이상하게 계속 안되고 있음. 원인을 알 수 없는 1로 인해 Output 사이즈가 자꾸 맞춰지지 않음.
    - 아ㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏ
h_bias를 제거함. 근데 다른 문제 생김.
->  0x167276600 <+928>: st1.4s { v0, v1 }, [x10], #32
    0x167276604 <+932>: add    x9, x22, #0x10
    0x167276608 <+936>: st1.4s { v2, v3 }, [x10]

    0x16727660c <+940>: add    x8, x8, #0x40
오버플로우남. 뒤질까
- 나는 멍청하다..
    - 진짜 멍청했다. 왜 backward를 하면 순서가 랜덤해질거라고 생각했지. 크기는 그ㅐㄹ..ㅇㅁ러민ㅇ러ㅏㅇ라어라ㅏㅓㅇ푸ㅠ유ㅜㅠㅜㅠㅜ유ㅜ류ㅜ유ㅜㅜㅠㅠㅜ 
 
may. 10. 2022 (b.insung)
- 아ㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏ
취중코디ㅣㅣㅣㅣㅣㅣㅣㅣㅣㅣㅣㅣㅇㅇㅇㅇㅇㅇㅇㅇ이ㅣ잉이이ㅣㅣ이이ㅣ이이이이ㅣㅣㅇㅇ
죄송합니다. 정말 진짜......하 이럴거면 술을 왜 마신거지
- 취중 코딩 중에 일단 한 짓 정리는 하자곱곱ㄱ
    - 일단 리스트 그대로 때려박았더니 리스트에서 overflow 발생. 
```
->  0x1798ca0dc <+424>: stp    q1, q0, [x9, #-0x20]
    0x1798ca0e0 <+428>: stp    q3, q2, [x9], #0x40
    0x1798ca0e4 <+432>: add    x22, x22, #0x8
    0x1798ca0e8 <+436>: cmp    x22, x26
Target 0: (python) stopped.
```

May. 9. 2022 (B. InSung)
- 나는 멍청하다... 어찌 라벨링 없이 구분을 할려고 했을까
- 하ㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏㅏ

May. 8. 2022 (B. InSung)
- Sunday Home session start.
- 이전에 짜둔 linear 코드는 SVM 을 거치지 않고 그냥 선형만 거치는 코드였다. 문제를 파악하고 새로운 SVM.py 코드 생성
    - RBM과 마찬가지로 메소드로 끌어와 작성할 예정

May. 6. 2022 (B. InSung)
- Friday Session start.

May. 4. 2022 (B. InSung)
- BBRBM 3개, GBRBM 3개 모두 쌓아서 Train Code 완성. AWS 내에서 테스트 할 예정
- 학습과정 중에 실수가 있었음. 실수로 인한 오류가 많아서 주피터 노트북으로 이전하여 부분적 디버깅으로 진행 예정.
- 현재 테스트 코드 작성 중. BBRBM 3개와 GBRBM을 어떻게 한번에 돌릴지를 궁리해야함. 이는 수업 끝나고 진행 예정.

May. 3. 2022. (B. InSung)
- Tuesday Session Start and want to sleep.
- 일단 새로운 모델로 서버에서 돌려 테스트 결과를 볼 예정.
- 논문에 따르면...
    - "Number of neurons for the first layer were varied within the set {180, 200, 250} and for the second layer, it varies in range {80, 100, 120}. All of them produce similar performance when the other hyper parameters are chosen accordingly."
    G., S., T., K.P. & V., K.K. Classification of ECG beats using deep belief network and active learning. Med Biol Eng Comput 56, 1887–1898 (2018). https://doi.org/10.1007/s11517-018-1815-2
        - 에서 레이어의 갯수를 정해주고 있다. 일단은 그에 맞게 출력을 해야하니 다시 해보자고...
- 일단 레이어 수는 맞췄음, 지금부터는 레이어수와 함께 그리고 loss 값을 토대로 입력을 해야함. 최종 결과 값은 AWS 내에서 돌려서 확인 후 출력 값으로 결정할 예정.
- BBRBM 레이어 성공. 추후 계획은 아래에
    - GBRBM을 설계하고 Testset으로 accuaracy 측정하여 결과값 보기.


wrong.py
# %%
import torch.distributions.distribution as D
from torch.utils.data import DataLoader
from torch.autograd import Variable
import torch.nn.functional as F
import matplotlib.pyplot as plt
import data.read_samples as rs
import torch.optim as optim
import torch.utils.data
import torch.nn as nn
import numpy as np
import datetime
import torch
import time
import os

print(datetime.datetime.now(), "model.py code start")

BATCH_SIZE = 10
EPOCH = 90
LEARNING_RATE = 0.2
ANNEALING_RATE = 0.999
VISIBLE_UNITS = [180, 200, 250]
HIDDEN_UNITS = [80, 100, 120]
K_FOLD = 1

# %%
device = torch.device('cuda')
print(torch.cuda.get_device_name(device))
os.environ['KMP_DUPLICATE_LIB_OK']='True'

# %%
class RBM(nn.Module): 
    with torch.cuda.device(0):
        def __init__(self, n_vis, n_hid, k, batch):
            super(RBM, self).__init__()
            self.W      = nn.Parameter(torch.randn(1, batch, device=device) * 1e-2)
            self.n_vis  = n_vis
            self.n_hid  = n_hid
            self.k      = k
            self.batch  = batch
            self.v_bias = nn.Parameter(torch.zeros(batch, device=device))
            self.h_bias = nn.Parameter(torch.zeros(batch, device=device))
        
        def sample_from_p(self, p):
            return F.relu(
                torch.sign(
                    p - Variable(torch.randn(p.size(), device=device))
                )
            ).to(device=device)

        ''' ISSUE PART '''
        def v_to_h(self, v):
            w = (self.W.clone())

            p_h = F.sigmoid(
                F.linear(v, w)
            ).to(device=device)

            sample_h = self.sample_from_p(p_h)
            return p_h, sample_h

        def h_to_v(self, h):
            w = self.W.t().clone()

            p_v = F.sigmoid(
                F.linear(h, w).to(device=device)
            ).to(device=device)

            sample_v = self.sample_from_p(p_v)
            return p_v, sample_v
        
        def forward(self, v):
            pre_h1, h1 = self.v_to_h(v)
            h_ = h1

            for _ in range(self.k):
                pre_v_, v_ = self.h_to_v(h_)
                pre_h_, h_ = self.v_to_h(v_)
            return v, v_
        
        def get_weight(self):
            return self.W

# %%
print("[MODL] Model main code is starting....")

print("[INFO] Read train data, cross-vaildation data and test data from median filtering code")
db1_sig, db1_label, db2_sig, db2_label, db3_sig, db3_label = rs.return_list()

# %%
train_dataloader = DataLoader(db1_sig + db2_sig,
                              batch_size=BATCH_SIZE,
                              num_workers=0, 
                              collate_fn=lambda x: x,
                              shuffle=True)
                              
test_dataloader = DataLoader(db2_sig + db3_sig,
                             batch_size=BATCH_SIZE,
                             num_workers=0, 
                             collate_fn=lambda x: x,
                             shuffle=True)

# %%
rbm_first = RBM(n_vis=VISIBLE_UNITS[0], n_hid=HIDDEN_UNITS[0], k=K_FOLD, batch=BATCH_SIZE).to(device=device)
rbm_second = RBM(n_vis=VISIBLE_UNITS[1], n_hid=HIDDEN_UNITS[1], k=K_FOLD, batch=BATCH_SIZE).to(device=device)
rbm_third = RBM(n_vis=VISIBLE_UNITS[2], n_hid=HIDDEN_UNITS[2], k=K_FOLD, batch=BATCH_SIZE).to(device=device)


first_train_op = optim.Adagrad(rbm_first.parameters(), 0.1)
second_train_op = optim.Adagrad(rbm_second.parameters(), 0.1)
third_train_op = optim.Adagrad(rbm_third.parameters(), 0.1)

gb_first_train_op = optim.Adagrad(rbm_first.parameters(), 0.1)
gb_second_train_op = optim.Adagrad(rbm_second.parameters(), 0.1)
gb_third_train_op = optim.Adagrad(rbm_third.parameters(), 0.1)

output_from_first = list()
output_from_second = list()
output_from_third = list()

omse_loss = list()
mse_loss = nn.MSELoss()

# %%
'''Train Part'''

loss_ = []
for epoch in range(EPOCH):
    start = time.time()
    '''First bbrbm'''
    for i, (data) in enumerate(train_dataloader):
        data = Variable(
                torch.tensor(data, dtype=torch.float32)
        ).uniform_(0, 1)
        
        sample_data = torch.bernoulli(data).view(-1, 10).to(device=device)

        # tensor binary
        vog_first, v1 = rbm_first(sample_data)
        omse_loss = mse_loss(vog_first, v1)

        first_train_op.zero_grad()
        first_train_op.step()
        omse_loss.backward()
    output_from_first.append(v1.tolist())
    print("1ST BBrbm_first Training loss for {0} epoch {1}\tEstimate time : {2}".format(epoch, omse_loss, time.time() - start))


# %%
output_from_first = torch.tensor(output_from_first)
print(output_from_first.size())

for epoch in range(EPOCH):
    '''Secnd bbrbm'''
    start = time.time()
    for _, (data) in enumerate(output_from_first): 
        data = Variable(
                torch.tensor(data, dtype=torch.float32)
        ).uniform_(0, 1)

        sample_data = torch.bernoulli(data).view(-1, 10).to(device=device)

        # tensor binary
        vog_second, v2 = rbm_second(sample_data)
        omse_loss = mse_loss(vog_second, v2)

        second_train_op.zero_grad()
        omse_loss.backward()
        second_train_op.step()

    end = time.time()
    output_from_second.append(v2.tolist())
    print("2ST BBrbm_first Training loss for {0} epoch {1}\tEstimate time : {2}".format(epoch, omse_loss, end - start))


# %%
output_from_second = torch.tensor(output_from_second)
for epoch in range(EPOCH):
    '''Third bbrbm'''
    for _, (data) in enumerate(output_from_second):
        start = time.time()
        data = Variable(
                torch.tensor(data, dtype=torch.float32)
        ).uniform_(0, 1)

        sample_data = torch.bernoulli(data).view(-1, 10).to(device=device)

        vog_third, v3 = rbm_third(sample_data)
        omse_loss = mse_loss(vog_third, v3)
        
        third_train_op.zero_grad()
        omse_loss.backward()
        third_train_op.step()
    
    end = time.time()
    output_from_third.append(v3)
    print("3ST BBrbm_first Training loss for {0} epoch {1}\tEstimate time : {2}".format(epoch, omse_loss, end - start))


# %%
# rbm_first = RBM(n_vis=VISIBLE_UNITS[0], n_hid=HIDDEN_UNITS[0], k=K_FOLD, batch=BATCH_SIZE).to(device=device)
# rbm_second = RBM(n_vis=VISIBLE_UNITS[1], n_hid=HIDDEN_UNITS[1], k=K_FOLD, batch=BATCH_SIZE).to(device=device)
# rbm_third = RBM(n_vis=VISIBLE_UNITS[2], n_hid=HIDDEN_UNITS[2], k=K_FOLD, batch=BATCH_SIZE).to(device=device)

output_from_first = list()
output_from_second = list()

omse_loss = list()
mse_loss = nn.MSELoss()
gaussian_std = torch.arange(1, 0, -0.1)

# %%
''' ** ISSUE PART ** '''

loss_ = []
for epoch in range(EPOCH):
    '''First gbrbm'''
    for i, (data) in enumerate(output_from_third):
        data = Variable(
                torch.tensor(data, dtype=torch.float32)
        ).uniform_(0, 1)
        
        sample_data = torch.normal(mean=data, std=gaussian_std).view(-1, 10)

        # tensor binary
        vog_first, v1 = rbm_first(sample_data)
        omse_loss = mse_loss(vog_first, v1)

        first_train_op.zero_grad()
        first_train_op.step()
        omse_loss.backward()
    output_from_first.append(v1.tolist())
    print("1ST GBrbm_first Training loss for {0} epoch {1}\tEstimate time : {2}".format(epoch, omse_loss))


# %%
output_from_first = torch.tensor(output_from_first)
output_from_third = list()
print(output_from_first.size())

for epoch in range(EPOCH):
    '''Secnd gbrbm'''
    start = time.time()
    for _, (data) in enumerate(output_from_first): 
        data = Variable(
                torch.tensor(data, dtype=torch.float32)
        ).uniform_(0, 1)

        sample_data = torch.normal(mean=data, std=gaussian_std).view(-1, 10)

        # tensor binary
        vog_second, v2 = rbm_second(sample_data)
        omse_loss = mse_loss(vog_second, v2)

        second_train_op.zero_grad()
        omse_loss.backward()
        second_train_op.step()

    end = time.time()
    output_from_second.append(v2.tolist())
    print("2ST BBrbm_first Training loss for {0} epoch {1}\tEstimate time : {2}".format(epoch, omse_loss, end - start))

# %%
output_from_second = torch.tensor(output_from_second)
for epoch in range(EPOCH):
    '''Third gbrbm'''
    for _, (data) in enumerate(output_from_second):
        start = time.time()
        data = Variable(
                torch.tensor(data, dtype=torch.float32)
        ).uniform_(0, 1)

        sample_data = torch.bernoulli(data).view(-1, 10)

        vog_third, v3 = rbm_third(sample_data)
        omse_loss = mse_loss(vog_third, v3)
        
        third_train_op.zero_grad()
        omse_loss.backward()
        third_train_op.step()
    
    end = time.time()
    output_from_third.append(v3)
    print("3ST BBrbm_first Training loss for {0} epoch {1}\tEstimate time : {2}".format(epoch, omse_loss, end - start))


# %%



